src/python/pants/backend/python/rules/resolve.py

```diff
@@ -0,0 +1,356 @@
+from pants.backend.python.target_types import (
+    PythonLibrary,
+    PythonRequirementLibrary,
+    PythonRequirementsField,
+)
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.addresses import Addresses
+from pants.engine.console import Console
+from pants.engine.fs import (
+    EMPTY_SNAPSHOT,
+    Digest,
+    DirectoryToMaterialize,
+    InputFilesContent,
+    MergeDigests,
+    PathGlobs,
+    Snapshot,
+    SnapshotSubset,
+    Workspace,
+)
+from pants.engine.goal import Goal, GoalSubsystem, Outputting
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import goal_rule, named_rule, rule
+from pants.engine.selectors import Get, MultiGet
+from pants.engine.target import Dependencies, Targets, TransitiveTargets
+from pants.option.custom_types import GlobExpansionConjunction
+from pants.option.global_options import GlobMatchErrorBehavior
+from pants.python.python_repos import PythonRepos
+from pants.python.python_setup import PythonSetup
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class Resolver(PexTool):
+    _entry_point = "resolver:main"
+    _module_names = ("resolver", "common")
+
+    pex: Pex
+
+
+@named_rule(desc="Create python resolver")
+async def create_resolver(pex_bin: PexBin) -> Resolver:
+    resolver_snapshot = await Get[Snapshot](InputFilesContent, Resolver.files())
+    init_injected = await Get[InitInjectedSnapshot](InjectInitRequest(snapshot=resolver_snapshot))
+    return Resolver(
+        pex=await Get[Pex](
+            PexRequest(
+                output_filename="bin/resolver.pex",
+                requirements=PexRequirements((pex_bin.requirement, "dataclasses==0.6")),
+                sources=init_injected.snapshot.digest,
+                entry_point=Resolver.entry_point(),
+                additional_args=("--no-strip-pex-env", "--unzip"),
+            )
+        )
+    )
+
+
+@dataclass(frozen=True)
+class PexCache:
+    snapshot: Snapshot = EMPTY_SNAPSHOT
+    path: str = "pex_root"
+
+
+@dataclass(frozen=True)
+class ResolveRequest:
+    requirements: PexRequirements
+    cache: PexCache = PexCache()
+    local: bool = False
+    requirer: Optional[str] = None
+
+
+@dataclass(frozen=True)
+class ResolveResult:
+    dependencies: DistributionDependencies = DistributionDependencies()
+    cache: PexCache = PexCache()
+
+
+@named_rule(desc="Resolve python requirements")
+async def resolve_requirements(
+    resolver: Resolver,
+    python_setup: PythonSetup,
+    python_repos: PythonRepos,
+    request: ResolveRequest,
+    subprocess_encoding_environment: SubprocessEncodingEnvironment,
+    log_level: LogLevel,
+) -> ResolveResult:
+    argv = list(PexDebug(log_level).iter_pex_args())
+
+    if python_setup.requirement_constraints is not None:
+        constraint_file_snapshot = await Get[Snapshot](
+            PathGlobs(
```

Comment:
We use this snippet twice, now. Might want to DRY as a method on `python_setup`. Because we can't factor out async functions, the method could return the `PathGlobs` object.